servant-server/src/Servant/Server/Internal.hs

```diff
@@ -703,3 +707,54 @@ instance (HasContextEntry context (NamedContext name subContext), HasServer subA
       subContext = descendIntoNamedContext (Proxy :: Proxy name) context
 
   hoistServerWithContext _ _ nt s = hoistServerWithContext (Proxy :: Proxy subApi) (Proxy :: Proxy subContext) nt s
+
+-------------------------------------------------------------------------------
+-- TypeError helpers
+-------------------------------------------------------------------------------
+
+#if MIN_VERSION_base(4,9,0)
+-- | This instance catches mistakes when there are non-saturated
+-- type applications on LHS of ':>'.
+--
+-- >>> serve (Proxy :: Proxy (Capture "foo" :> Get '[JSON] Int)) (error "...")
+-- ...
+-- ...Expected something of kind Symbol or *, got: k -> l on the LHS of ':>'.
+-- ...Maybe you haven't applied enough arguments to
+-- ...Capture "foo"
+-- ...
+instance TypeError
+    ('Text "Expected something of kind Symbol or *, got: k -> l on the LHS of ':>'."
```

Comment:
Don't you want `'ShowType k :<>: 'Text "->" :<>: 'ShowType l` instead of `k -> l`?

Reply:
We could, but then we'd need `{-# LANGUAGE TypeInType #-}`, and then i'd rather use `Type`, and not `*` everywhere we have.

The error message doesn't get much better:
```diff
-Expected something of kind Symbol or *, got: k -> l on the LHS of ':>'.
+Expected something of kind Symbol or *, got: * -> * on the LHS of ':>'.
```

My plan is to have 0.13 with GHC-7.8 support, but hopefully for next release GHC-8.4 will be out, so we can drop support of GHC-7.8 and GHC-7.10, and clean up a lot of things. This too. Does this sound as a good plan?

Reply:
Yeah, I guess. The problem I have with `k -> l` is the same as with `a -> b` below: `k` and `l` are not any of the user kind variables, so it can confuse them actually more than `* -> *`.

Anyway, it's ok to clean this up later, and the error message is a lot better already :)